### ISSUE TITLE
New tier-2 target for `wasm32-wasi-Preview 2`

### ISSUE THREAD

**Issue by yoshuawuyts** [rust-lang/compiler-team#594]
# Proposal

In 2019 the first version of the [WASI spec](https://wasi.dev) was released (what we now call "WASI Preview 1"), and support was added to rustc as the tier-2 `wasm32-wasi` target.
The folks working on WASI have iterated a lot on it since, resulting in major changes to the spec. Among other things, there is now a separate notion of "the component model", the way different WASI components bind to each other has changed, and WASI APIs are now defined in a new IDL language called [wit] [^overview]. This new version of the spec is called "WASI Preview 2", and is expected to be finalized later this year. It's also largely incompatible with the existing WASI Preview 1 spec.

[wit]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md

[^overview]: For anyone interested in understanding these changes in more depth, see Luke Wagner's talk on the component model ([YouTube](https://github.com/bytecodealliance/wit-bindgen))

Together with @sunfishcode and @alexcrichton I've been working to chart a path to support WASI Preview 2 in Rust. We believe the time for this is right, since the Preview 2 spec is nearly finalized, support in runtimes has begun (e.g. wasmtime), and there is industry demand for this (e.g. Fastly, Microsoft - our respective employers).

We want to be careful not to break existing users of Preview 1, so we believe the best way to add Preview 2 support is to introduce a new target triple: `wasm32-wasi-Preview 2`.
By introducing a new triple we can gradually implement support for Preview 2, and we provide users of Preview 1 the opportunity to upgrade to Preview 2 at their own pace. The implementation of Preview 2 in `rust-lang/rust` should mostly touch the stdlib APIs. Packaging the compiler's output into WASI Preview 2-compatible components is something which we expect for the time being to remain handled through external tools such as [`cargo-component`].

[`cargo-component`]: https://github.com/bytecodealliance/cargo-component

After the Preview 2 spec has been ratified, the plan by the WASI spec authors is to begin work on a Preview 3 version of the spec with first-class support for async primitives [^async]. And once that's done, they expect to work towards a stable "WASI 1.0" release ([presentation][roadmap]). To support WASI Preview 3 in Rust, we'll likely want to repeat the process for Preview 2: introduce a `wasm32-wasi-Preview 3` triple, and then work to support that in the compiler and stdlib. To add WASI 1.0 support however, we'll likely want to reuse the existing `wasm32-wasi` target which is currently in use by WASI Preview 1. But that's at least several years out, and worth discussing in its own dedicated change proposal.

[^async]: In terms of deltas, the difference between "Preview 1" and "Preview 2" is quite big because most of the primitives have been redesigned. In comparison: the difference between "Preview 2" and "Preview 3" should be relatively small; it adds support for "real" futures and streams on top of the Preview 2 model. And finally the difference between "Preview 3" and "1.0" should be even smaller than that; mostly serving as one last opportunity to fix whatever papercuts remain in Preview 3.

[roadmap]: https://github.com/WebAssembly/meetings/blob/main/wasi/2023/presentations/2023-02-09-gohman-wasi-roadmap.pdf

[`wasi` crate]: https://github.com/sunfishcode/rust-wasi/tree/sunfishcode/experiment

# Mentors or Reviewers

[tbd]

# Process

The main points of the [Major Change Process][MCP] are as follows:

* [x] File an issue describing the proposal.
* [ ] A compiler team member or contributor who is knowledgeable in the area can **second** by writing `@rustbot second`.
    * Finding a "second" suffices for internal changes. If however, you are proposing a new public-facing feature, such as a `-C flag`, then full team check-off is required.
    * Compiler team members can initiate a check-off via `@rfcbot fcp merge` on either the MCP or the PR.
* [ ] Once an MCP is seconded, the Final Comment Period begins. If no objections are raised after 10 days, the MCP is considered **approved**.

You can read [more about Major Change Proposals on forge][MCP].

[MCP]: https://forge.rust-lang.org/compiler/mcp.html

# Comments

**This issue is not meant to be used for technical discussion. There is a Zulip stream for that. Use this issue to leave procedural comments, such as volunteering to review, indicating that you second the proposal (or third, etc), or raising a concern that you would like to be addressed.**

**Comment by Zhangyx24** [comment]
Is preview2 still described using wit？

**Comment by jeffparsons** [comment]
> Is preview2 still described using wit?

Yep. The component model, on top of which WASI builds, uses the wit format to describe interfaces. (I.e. no more witx, and wit is now on a standards track.)
